src/jit/importer.cpp

```diff
@@ -2773,57 +2783,177 @@ GenTreePtr      Compiler::impInitializeArrayIntrinsic(CORINFO_SIG_INFO * sig)
+                beginArgs = numArgsArg->Rest();
+        }
+        else
+        {
+            beginArgs = args;
+
+            for (GenTreeArgList* arg = args; arg != nullptr; arg = arg->Rest())
+            {
+                bool isSecondToLast = (arg->Rest() != nullptr) && (arg->Rest()->Rest() == nullptr);
+
+                if (isSecondToLast)
+                {
+                    numArgsArg = arg;
+                    endArgs = arg;
+                    break;
+                }
+            }
+        }
+
+        //
+        // The number of arguments should be a constant between 1 and 64. The rank can't be 0
+        // so at least one length must be present and the rank can't exceed 32 so there can
+        // be at most 64 arguments - 32 lengths and 32 lower bounds.
+        //
+
+        if ((beginArgs == nullptr) ||
+            (numArgsArg == nullptr) ||
+            (numArgsArg->Current()->OperGet() != GT_CNS_INT) ||
+            (numArgsArg->Current()->AsIntCon()->IconValue() < 1) ||
+            (numArgsArg->Current()->AsIntCon()->IconValue() > 64))
+        {
+            return nullptr;
+        }
+
+        unsigned numArgs = static_cast<unsigned>(numArgsArg->Current()->AsIntCon()->IconValue());
+        bool lowerBoundsSpecified;
+
+        if (numArgs == rank * 2)
+        {
+            lowerBoundsSpecified = true;
+        }
+        else if (numArgs == rank)
+        {
+            lowerBoundsSpecified = false;
+
+            //
+            // If the rank is 1 and a lower bound isn't specified then the runtime creates
+            // a SDArray. Note that even if a lower bound is specified it can be 0 and then
+            // we get a SDArray as well, see the for loop below.
+            //
+
+            if (rank == 1)
+                isMDArray = false;
+        }
+        else
+        {
+            return nullptr;
+        }
+
+        //
+        // The rank is known to be at least 1 so we can start with numElements being 1
+        // to avoid the need to special case the first dimension.
+        //
+
+        numElements = S_UINT32(1);
+
+        for (GenTreeArgList* arg = beginArgs; arg != endArgs; arg = arg->Rest())
+        {
+            if (lowerBoundsSpecified)
+            {
+                //
+                // In general lower bounds can be ignored because they're not needed to
+                // calculate the total number of elements. But for single dimensional arrays
+                // we need to know if the lower bound is 0 because in this case the runtime
+                // creates a SDArray and this affects the way the array data offset is calculated.
+                //
+
+                if (rank == 1)
+                {
+                    GenTree* lowerBoundNode = arg->Current();
+
+                    if (lowerBoundNode->OperGet() != GT_CNS_INT)
+                        return nullptr;
+
+                    if (lowerBoundNode->AsIntCon()->IconValue() == 0)
```

Comment:
Minor suggestion: You could simply check `if (lowerBoundNode->IsZero())` here, and omit the check for non-constant, which will be taken care of below - makes the code a little simpler to understand.